### PR TITLE
Add new reporting intervals for the sales log with common sales periods....

### PR DESCRIPTION
**Pull request by 1bigidea**
... Provide filters so that specified periods and calculations can be overridden or extended.

### CODE REVIEW

**Comment by 1bigidea** [comment]
This should completely unducked. Thanks @JeffPyeBrook for the pointers (I even improved on the process by learning about a soft Reset! This could be improved with some unit testing but one step at a time!

**Comment by JustinSainton** [comment]
:+1:  FANTASTIC work so far.  Ready for #462? :100: 
 
A few cursory notes:
* Prefix filter names with `wpsc`
* Big plus one on unit tests.  Would love to merge with unit tests.  Eventually, we'll work our way towards _only_ merging PRs with unit tests.
* Some minor whitespace issues.

File: wpsc-admin/includes/purchase-log-list-table-class.php [hunk 1]
```diff
@@ -432,9 +613,55 @@ public function months_dropdown( $post_type = '' ) {
...
+			4 => __('Last Week', 'wpsc'),
+			5 => __('This Month', 'wpsc'),
+			6 => __('Last Month', 'wpsc'),
+			7 => __('This Quarter', 'wpsc'),
+			8 => __('Prior 3 Months', 'wpsc'),
+			9 => __('This Year', 'wpsc'),
+			10 => __('Last Year', 'wpsc'),
+		) );
+
+		echo '<option disabled="disabled">---------</option>';
```
Comment (JustinSainton) [inline wpsc-admin/includes/purchase-log-list-table-class.php]: Probably need something a bit more descriptive and internationalized here.

File: wpsc-admin/includes/purchase-log-list-table-class.php [hunk 2]
```diff
@@ -432,9 +613,55 @@ public function months_dropdown( $post_type = '' ) {
...
+		/**
+		 * Filter the available special periods on the purchase log listing screen.
+		 *
+		 * Can Used to remove periods or add new period definitions {@see purchase_log_predefined_periods_}
+		 *
+		 * @since 4.1.0
+		 *
+		 * @param array array() The periods currently defined.
+		 */
+		$periods = apply_filters( 'purchase_log_special_periods', array(
```
Comment (JustinSainton) [inline wpsc-admin/includes/purchase-log-list-table-class.php]: Filters like this should be prefixed with wpsc_

File: wpsc-admin/includes/purchase-log-list-table-class.php [hunk 3]
```diff
@@ -432,9 +613,55 @@ public function months_dropdown( $post_type = '' ) {
 			</select>
 			<?php
 			submit_button( _x( 'Filter', 'extra navigation in purchase log page', 'wpsc' ), 'secondary', false, false, array( 'id' => 'post-query-submit' ) );
+//			submit_button( _x( 'Export', 'extra navigation in purchase log page', 'wpsc' ), 'secondary', false, false, array( 'id' => 'wpec-export-log', name => "request-csv" ) );
```
Comment (JustinSainton) [inline wpsc-admin/includes/purchase-log-list-table-class.php]: Commented out lines of code are not necessary.

File: wpsc-admin/includes/purchase-log-list-table-class.php [hunk 4]
```diff
@@ -193,6 +211,166 @@ public function prepare_items() {
...
+				 * Return a custom date query for custom period_flag's.
+				 *
+				 * This filter extends the functionality allowing for custom periods if a new value
+				 * is passed via $_REQUEST['m']. {@see 'purchase_log_special_periods'}.
+				 *
+				 * @since 4.1.0
+				 *
+				 * @param array         Empty array to be filled with a valid query {@see WP_Date_Query}.
+				 */
+				$date_query = apply_filters( 'purchase_log_predefined_periods_'. $period_flag, array() );
```
Comment (JustinSainton) [inline wpsc-admin/includes/purchase-log-list-table-class.php]: Little bit of whitespace here (with the concatenation)

**Comment by 1bigidea** [comment]
besides cleanup (as noted above) - the next stop here is to remove that damned Export link at the bottom of the page and make it a first-class citizen at the top - right next to filter - so you can control the number of entries in the csv file itself (rather than having it choke itself in memory).

On a related thought, what would you think of a wp-cli command for exporting the CSV.

**Comment by JustinSainton** [comment]
:+1: on WP-CLI.  See #193.  Would love an entire WP-CLI suite for WPeC, actually.  Generating products, customers, orders, importing, exporting - that would all be *rad*.

**Comment by leewillis77** [comment]
Related: https://github.com/leewillis77/WP-CLI-WPEC-Command

**Comment by JustinSainton** [comment]
Oh yes, I forgot about that, @leewillis77!  Can we import that as a branch to our main repo, so committers like you, @1bigidea and myself can all contribute to the branch?

**Comment by JustinSainton** [comment]
Because _that_ would be awesome.

**Comment by leewillis77** [comment]
Currently it's a standalone thing, should probably be part of the main plugin (Like EDD does it). I'll integrate and raise a separate PR...

**Comment by JustinSainton** [comment]
Affirmative :+1:

**Comment by 1bigidea** [comment]
Updated per @JustinSainton notes

**Comment by JustinSainton** [comment]
@1bigidea Now...how do you feel about unit tests? :)

**Comment by 1bigidea** [comment]
Grumble, Grumble, Grumble 
Kicking dirt on home plate
Shooting evil looks at the umpire while headed back to the dugout

I’ll probably be asking for some batting tips before the next inning.

(Seasonal carping)



Every successful WordPress project starts with...
One Big Idea

(360) 746-2101                         transom@1bigidea.com
(815) 642-0258 efax 
gChat: transom@1bigidea.com    SKYPE: tomransom
Twitter: wp_site_manager                     www.linkedin.com/in/onebigidea


On April 29, 2015 at 12:58:18 PM, Justin Sainton (notifications@github.com) wrote:

@1bigidea Now...how do you feel about unit tests? :)

**Comment by JustinSainton** [comment]
haha, we can merge without for now. I probably need to put up something in the Wiki regarding unit tests.  Will likely make it required after 4.0 ships.

**Comment by 1bigidea** [comment]
I’ll make you a deal - rather than doing unit test for this POS class.

I will provide Unit tests for the rewritten WPSC_Purchase_Log_Query class as well as what little remains in this class.

This has two benefits - 1) I can postpone the problem, and 2) we have tests for a re-factored version of this blech of code.



Every successful WordPress project starts with...
One Big Idea

(360) 746-2101                         transom@1bigidea.com
(815) 642-0258 efax 
gChat: transom@1bigidea.com    SKYPE: tomransom
Twitter: wp_site_manager                     www.linkedin.com/in/onebigidea


On April 29, 2015 at 12:58:18 PM, Justin Sainton (notifications@github.com) wrote:

@1bigidea Now...how do you feel about unit tests? :)

**Comment by JustinSainton** [comment]
that'll do :P

**Pull request merged**